File: frontend/src/api/clientPortal.ts
import { apiClient } from './client'

export interface ClientPortalAccess {
  id: string
  userId: string
  projectId: string
  isActive: boolean
  canViewBudget: boolean
  canViewDocuments: boolean
  canSubmitFeedback: boolean
  lastAccessedAt?: string
  createdAt: string
}

export interface ClientPortalAuthRequest {
  email: string
  accessToken: string
}

export interface ClientPortalAuthResponse {
  accessToken: string
  tokenType: string
  userEmail: string
  userFullName?: string
  projectId: string
  projectName: string
  canViewBudget: boolean
  canViewDocuments: boolean
  canSubmitFeedback: boolean
}

export interface ClientPortalProject {
  id: string
  name: string
  description?: string
  address?: string
  startDate?: string
  estimatedEndDate?: string
  status: string
  website?: string
  imageUrl?: string
  locationLat?: number
  locationLng?: number
  locationAddress?: string
  daysRemaining?: number
  budgetVisibleToClients: boolean
  milestoneTrackingEnabled: boolean
}

export interface ClientPortalProgress {
  totalTasks: number
  completedTasks: number
  inProgressTasks: number
  completionPercentage: number
  totalMilestones: number
  completedMilestones: number
  upcomingMilestones: number
  onTrack: boolean
  lastUpdated: string
}

export interface ClientPortalBudgetSummary {
  totalBudgeted: string
  totalActual: string
  totalVariance: string
  variancePercentage: number
  approvedChangeOrders: number
  totalChangeOrderAmount: string
}

export interface ClientPortalPhoto {
  id: string
  filename: string
  fileType?: string
  fileSize?: number
  storagePath: string
  uploadedAt: string
  entityType: string
  entityId: string
  areaName?: string
  floorNumber?: number
}

export interface ClientPortalDocument {
  id: string
  filename: string
  fileType?: string
  fileSize?: number
  storagePath: string
  uploadedAt: string
  entityType: string
  category?: string
}

export interface ClientPortalFeedbackCreate {
  subject: string
  content: string
  entityType?: string
  entityId?: string
}

export interface ClientPortalFeedback {
  id: string
  subject: string
  content: string
  entityType: string
  entityId: string
  submittedBy: string
  submittedAt: string
  status: string
}

export interface ClientPortalMilestone {
  id: string
  title: string
  description?: string
  targetDate?: string
  completedAt?: string
  status: string
  completionPercentage: number
  isOverdue: boolean
}

export interface TimelineEvent {
  id: string
  date: string
  title: string
  description?: string
  eventType: string
  entityId?: string
  entityType?: string
  userName?: string
  metadata?: Record<string, unknown>
}

export interface ProgressMetrics {
  overallPercentage: number
  inspectionsCompleted: number
  inspectionsTotal: number
  equipmentSubmitted: number
  equipmentTotal: number
  materialsSubmitted: number
  materialsTotal: number
  checklistsCompleted: number
  checklistsTotal: number
}

export interface TeamStats {
  totalMembers: number
  activeMembers: number
  roles: Record<string, number>
}

export interface ProjectStats {
  totalInspections: number
  pendingInspections: number
  totalEquipment: number
  totalMaterials: number
  totalMeetings: number
  openFindings: number
  daysRemaining?: number
  daysElapsed?: number
}

export interface ProjectOverview {
  projectId: string
  projectName: string
  projectStatus: string
  progress: ProgressMetrics
  timeline: TimelineEvent[]
  teamStats: TeamStats
  stats: ProjectStats
  lastUpdated: string
  locationLat?: number
  locationLng?: number
  locationAddress?: string
}

export interface ProjectMember {
  id: string
  userId: string
  user: {
    id: string
    email: string
    fullName: string
    isActive: boolean
  }
  role: string
  addedAt: string
}

export interface Project {
  id: string
  name: string
  description?: string
  address?: string
  startDate?: string
  estimatedEndDate?: string
  status: string
  dailySummaryEnabled: boolean
  notificationDigestIntervalHours: number
  lastDigestSentAt?: string
  createdAt: string
  updatedAt: string
  members: ProjectMember[]
  website?: string
  imageUrl?: string
  locationLat?: number
  locationLng?: number
  locationAddress?: string
  daysRemaining?: number
}

export interface FileResponse {
  id: string
  projectId: string
  entityType: string
  entityId: string
  filename: string
  fileType?: string
  fileSize?: number
  storagePath: string
  uploadedAt: string
  uploadedBy?: {
    id: string
    email: string
    fullName: string
    isActive: boolean
  }
}

export interface DiscussionAuthor {
  id: string
  fullName: string
  email: string
}

export interface Discussion {
  id: string
  projectId: string
  entityType: string
  entityId: string
  authorId: string
  parentId?: string
  content: string
  createdAt: string
  updatedAt: string
  author?: DiscussionAuthor
  replies: Discussion[]
}

export interface DiscussionCreate {
  entityType: string
  entityId: string
  content: string
  parentId?: string
}

export const clientPortalApi = {
  authenticate: async (data: ClientPortalAuthRequest): Promise<ClientPortalAuthResponse> => {
    const response = await apiClient.post<ClientPortalAuthResponse>('/client-portal/auth', {
      email: data.email,
      access_token: data.accessToken
    })
    return response.data
  },

  healthCheck: async (): Promise<{ status: string; service: string }> => {
    const response = await apiClient.get<{ status: string; service: string }>('/client-portal/health')
    return response.data
  },

  listProjects: async (): Promise<Project[]> => {
    const response = await apiClient.get<Project[]>('/client-portal/projects')
    return response.data
  },

  getProject: async (projectId: string): Promise<Project> => {
    const response = await apiClient.get<Project>(`/client-portal/projects/${projectId}`)
    return response.data
  },

  getProjectOverview: async (projectId: string): Promise<ProjectOverview> => {
    const response = await apiClient.get<ProjectOverview>(`/client-portal/projects/${projectId}/overview`)
    return response.data
  },

  getProjectProgress: async (projectId: string): Promise<ProgressMetrics> => {
    const response = await apiClient.get<ProgressMetrics>(`/client-portal/projects/${projectId}/progress`)
    return response.data
  },

  listPhotos: async (projectId: string, limit = 50, offset = 0): Promise<FileResponse[]> => {
    const response = await apiClient.get<FileResponse[]>(
      `/client-portal/projects/${projectId}/photos`,
      { params: { limit, offset } }
    )
    return response.data
  },

  listDocuments: async (
    projectId: string,
    entityType?: string,
    limit = 50,
    offset = 0
  ): Promise<FileResponse[]> => {
    const response = await apiClient.get<FileResponse[]>(
      `/client-portal/projects/${projectId}/documents`,
      { params: { entity_type: entityType, limit, offset } }
    )
    return response.data
  },

  listFeedback: async (projectId: string): Promise<Discussion[]> => {
    const response = await apiClient.get<Discussion[]>(`/client-portal/projects/${projectId}/feedback`)
    return response.data
  },

  createFeedback: async (projectId: string, data: DiscussionCreate): Promise<Discussion> => {
    const response = await apiClient.post<Discussion>(
      `/client-portal/projects/${projectId}/feedback`,
      data
    )
    return response.data
  },
}

File: backend/app/schemas/client_portal.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.core.validators import (
    MAX_DESCRIPTION_LENGTH,
    CamelCaseModel,
    sanitize_string,
)


class ClientPortalAccessResponse(CamelCaseModel):
    id: UUID
    user_id: UUID
    project_id: UUID
    is_active: bool
    can_view_budget: bool
    can_view_documents: bool
    can_submit_feedback: bool
    last_accessed_at: Optional[datetime] = None
    created_at: datetime


class ClientPortalAuthRequest(BaseModel):
    email: EmailStr
    access_token: str = Field(min_length=1, max_length=255)


class ClientPortalAuthResponse(CamelCaseModel):
    access_token: str
    token_type: str = "bearer"
    user_email: str
    user_full_name: Optional[str] = None
    project_id: UUID
    project_name: str
    can_view_budget: bool
    can_view_documents: bool
    can_submit_feedback: bool


class ClientPortalProjectResponse(CamelCaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    address: Optional[str] = None
    start_date: Optional[date] = None
    estimated_end_date: Optional[date] = None
    status: str
    website: Optional[str] = None
    image_url: Optional[str] = None
    location_lat: Optional[float] = None
    location_lng: Optional[float] = None
    location_address: Optional[str] = None
    days_remaining: Optional[int] = None
    budget_visible_to_clients: bool = False
    milestone_tracking_enabled: bool = True


class ClientPortalProgressResponse(CamelCaseModel):
    total_tasks: int = 0
    completed_tasks: int = 0
    in_progress_tasks: int = 0
    completion_percentage: float = 0.0
    total_milestones: int = 0
    completed_milestones: int = 0
    upcoming_milestones: int = 0
    on_track: bool = True
    last_updated: datetime


class ClientPortalBudgetSummaryResponse(CamelCaseModel):
    total_budgeted: Decimal
    total_actual: Decimal
    total_variance: Decimal
    variance_percentage: float
    approved_change_orders: int
    total_change_order_amount: Decimal


class ClientPortalPhotoResponse(CamelCaseModel):
    id: UUID
    filename: str
    file_type: Optional[str] = None
    file_size: Optional[int] = None
    storage_path: str
    uploaded_at: datetime
    entity_type: str
    entity_id: UUID
    area_name: Optional[str] = None
    floor_number: Optional[int] = None


class ClientPortalDocumentResponse(CamelCaseModel):
    id: UUID
    filename: str
    file_type: Optional[str] = None
    file_size: Optional[int] = None
    storage_path: str
    uploaded_at: datetime
    entity_type: str
    category: Optional[str] = None


class ClientPortalFeedbackCreate(BaseModel):
    subject: str = Field(min_length=2, max_length=200)
    content: str = Field(min_length=1, max_length=MAX_DESCRIPTION_LENGTH)
    entity_type: Optional[str] = Field(default="project", max_length=50)
    entity_id: Optional[UUID] = None

    @field_validator('subject', 'content', mode='before')
    @classmethod
    def sanitize_text(cls, v: Optional[str]) -> Optional[str]:
        return sanitize_string(v)


class ClientPortalFeedbackResponse(CamelCaseModel):
    id: UUID
    subject: str
    content: str
    entity_type: str
    entity_id: UUID
    submitted_by: str
    submitted_at: datetime
    status: str = "pending"


class ClientPortalMilestoneResponse(CamelCaseModel):
    id: UUID
    title: str
    description: Optional[str] = None
    target_date: Optional[date] = None
    completed_at: Optional[datetime] = None
    status: str
    completion_percentage: float = 0.0
    is_overdue: bool = False


class ClientPortalOverviewResponse(CamelCaseModel):
    project: ClientPortalProjectResponse
    progress: ClientPortalProgressResponse
    budget_summary: Optional[ClientPortalBudgetSummaryResponse] = None
    recent_photos: list[ClientPortalPhotoResponse] = []
    upcoming_milestones: list[ClientPortalMilestoneResponse] = []
